packages/flutter/lib/src/material/carousel.dart

```diff
@@ -543,25 +583,32 @@ class _SliverFixedExtentCarousel extends SliverMultiBoxAdaptorWidget {
     required super.delegate,
     required this.minExtent,
     required this.itemExtent,
+    required this.controller,
   });
 
   final double itemExtent;
   final double minExtent;
 
+  /// Called whenever the current item index changes, either due to
+  /// user scrolling or programmatic navigation.
+  final CarouselController controller;
+
   @override
   RenderSliverFixedExtentBoxAdaptor createRenderObject(BuildContext context) {
     final SliverMultiBoxAdaptorElement element = context as SliverMultiBoxAdaptorElement;
     return _RenderSliverFixedExtentCarousel(
       childManager: element,
       minExtent: minExtent,
       maxExtent: itemExtent,
+      controller: controller,
     );
   }
 
   @override
   void updateRenderObject(BuildContext context, _RenderSliverFixedExtentCarousel renderObject) {
     renderObject.maxExtent = itemExtent;
     renderObject.minExtent = minExtent;
+    renderObject.controller = controller;
```

Comment:
Passing the controller down to the render objects is not a good approach IMO. 
By passing in a controller, the RenderObject is now directly mutating state on an object from a higher layer (`controller._currentIndex = newIndex)`. We really should not do this.

Controllers belong in the widget/state layer. They are meant to coordinate actions and manage state that is exposed to the developer.
RenderObjects are in the rendering layer, and should be focused only on layout and painting based on their given constraints and properties.